Kryo is not ThreadSafe

Kryo is not ThreadSafe, mybe CacheRevision should use ThreadLocal<Kryo>
See https://github.com/EsotericSoftware/kryo#threading

We get this after updating to 1.1.3:

`2017-12-08T10:11:49,517 ERROR [main] (Main.java:59) - Fatal error: null
java.lang.ClassCastException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:1.8.0_45]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:1.8.0_45]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:1.8.0_45]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422) ~[?:1.8.0_45]
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598) ~[?:1.8.0_45]
	at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677) ~[?:1.8.0_45]
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735) ~[?:1.8.0_45]
	at java.util.concurrent.ConcurrentHashMap.forEachEntry(ConcurrentHashMap.java:4195) ~[?:1.8.0_45]
	at svnserver.repository.mapping.RepositoryListMapping.initRevisions(RepositoryListMapping.java:57) ~[git-as-svn.jar:?]
	at svnserver.server.SvnServer.<init>(SvnServer.java:125) ~[git-as-svn.jar:?]
	at svnserver.server.Main.main(Main.java:48) [git-as-svn.jar:?]
Caused by: java.lang.ClassCastException: org.eclipse.jgit.lib.ObjectId cannot be cast to svnserver.repository.git.cache.CacheChange
	at svnserver.repository.git.GitRepository.lambda$loadRevisionInfo$1(GitRepository.java:382) ~[git-as-svn.jar:?]
	at svnserver.repository.git.GitRepository$$Lambda$29/1321056859.apply(Unknown Source) ~[?:?]
	at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1892) ~[?:1.8.0_45]
	at svnserver.repository.git.GitRepository.loadRevisionInfo(GitRepository.java:379) ~[git-as-svn.jar:?]
	at svnserver.repository.git.GitRepository.loadRevisions(GitRepository.java:209) ~[git-as-svn.jar:?]
	at svnserver.repository.git.GitRepository.updateRevisions(GitRepository.java:352) ~[git-as-svn.jar:?]
	at svnserver.repository.mapping.RepositoryListMapping.lambda$initRevisions$0(RepositoryListMapping.java:59) ~[git-as-svn.jar:?]
	at svnserver.repository.mapping.RepositoryListMapping$$Lambda$27/577127077.accept(Unknown Source) ~[?:?]
	at java.util.concurrent.ConcurrentHashMap$ForEachEntryTask.compute(ConcurrentHashMap.java:4977) ~[?:1.8.0_45]
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731) ~[?:1.8.0_45]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) ~[?:1.8.0_45]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) ~[?:1.8.0_45]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1689) ~[?:1.8.0_45]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) ~[?:1.8.0_45]
Exception in thread "main" java.lang.ClassCastException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598)
	at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
	at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735)
	at java.util.concurrent.ConcurrentHashMap.forEachEntry(ConcurrentHashMap.java:4195)
	at svnserver.repository.mapping.RepositoryListMapping.initRevisions(RepositoryListMapping.java:57)
	at svnserver.server.SvnServer.<init>(SvnServer.java:125)
	at svnserver.server.Main.main(Main.java:48)
Caused by: java.lang.ClassCastException: org.eclipse.jgit.lib.ObjectId cannot be cast to svnserver.repository.git.cache.CacheChange
	at svnserver.repository.git.GitRepository.lambda$loadRevisionInfo$1(GitRepository.java:382)
	at svnserver.repository.git.GitRepository$$Lambda$29/1321056859.apply(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1892)
	at svnserver.repository.git.GitRepository.loadRevisionInfo(GitRepository.java:379)
	at svnserver.repository.git.GitRepository.loadRevisions(GitRepository.java:209)
	at svnserver.repository.git.GitRepository.updateRevisions(GitRepository.java:352)
	at svnserver.repository.mapping.RepositoryListMapping.lambda$initRevisions$0(RepositoryListMapping.java:59)
	at svnserver.repository.mapping.RepositoryListMapping$$Lambda$27/577127077.accept(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap$ForEachEntryTask.compute(ConcurrentHashMap.java:4977)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1689)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
`

Either our cache is corrupt (but I failed to understand what git-as-svn version could write data that way) or git-as-svn kryo usage still isn't thread safe enough.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,4 +1,9 @@
 # Changes
+
+## 1.1.4
+
+ * Upgrade Kryo to 4.0.1
+
 ## 1.1.3
 
  * Fix ISO 8601 date formatting.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,7 +2,8 @@
 
 ## 1.1.4
 
- * Upgrade Kryo to 4.0.1
+ * Upgrade Kryo to 4.0.1 #121
+ * Add option to disable parallel repository indexing on startup #121
 
 ## 1.1.3
 

--- a/build.gradle
+++ b/build.gradle
@@ -111,7 +111,7 @@ dependencies {
   compile "org.eclipse.jetty:jetty-server:9.3.7.v20160115"
   compile "org.eclipse.jetty:jetty-servlet:9.3.7.v20160115"
   compile "org.gitlab:java-gitlab-api:1.2.3"
-  compile "com.esotericsoftware:kryo:3.0.3"
+  compile "com.esotericsoftware:kryo:4.0.1"
   compile "org.bitbucket.b_c:jose4j:0.4.4"
   compile "com.google.guava:guava:19.0"
   compile "com.kohlschutter.junixsocket:junixsocket-native-common:2.0.4"

--- a/cfg/config-gitlab.example
+++ b/cfg/config-gitlab.example
@@ -1,6 +1,7 @@
 !config:
 realm: Example realm
 compressionEnabled: true
+parallelIndexing: true
 
 # Use GitLab repositories
 repositoryMapping: !gitlabMapping

--- a/cfg/config-local.example
+++ b/cfg/config-local.example
@@ -19,6 +19,10 @@
 #
 # compressionEnabled: true
 
+# Use parallel repository indexing on startup
+#
+# parallelIndexing: true
+
 # Set cache location
 cacheConfig: !persistentCache
   path: ../build/git-as-svn.mapdb

--- a/cfg/config.example
+++ b/cfg/config.example
@@ -19,6 +19,10 @@
 #
 # compressionEnabled: true
 
+# Use parallel repository indexing on startup
+#
+# parallelIndexing: true
+
 # Set cache location
 cacheConfig: !persistentCache
   path: /var/cache/git-as-svn/git-as-svn.mapdb

--- a/src/main/java/svnserver/config/Config.java
+++ b/src/main/java/svnserver/config/Config.java
@@ -48,6 +48,7 @@ public final class Config {
   private boolean reuseAddress = false;
   private boolean compressionEnabled = true;
   private long shutdownTimeout = TimeUnit.SECONDS.toMillis(5);
+  private boolean parallelIndexing = true;
 
   @SuppressWarnings("UnusedDeclaration")
   public Config() {
@@ -110,6 +111,10 @@ public final class Config {
   @NotNull
   public List<SharedConfig> getShared() {
     return shared;
+  }
+
+  public boolean canUseParallelIndexing() {
+    return parallelIndexing;
   }
 
   public boolean isCompressionEnabled() {

--- a/src/main/java/svnserver/ext/gitlab/mapping/GitLabMapping.java
+++ b/src/main/java/svnserver/ext/gitlab/mapping/GitLabMapping.java
@@ -24,9 +24,9 @@ import svnserver.repository.mapping.RepositoryListMapping;
 
 import java.io.File;
 import java.io.IOException;
-import java.util.Map;
-import java.util.NavigableMap;
+import java.util.*;
 import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.stream.Collectors;
 
 /**
  * Simple repository mapping by predefined list.
@@ -65,10 +65,9 @@ public class GitLabMapping implements VcsRepositoryMapping {
   }
 
   @Override
-  public void initRevisions() throws IOException, SVNException {
-    for (Map.Entry<String, GitLabProject> entry : mapping.entrySet()) {
-      entry.getValue().initRevisions();
-    }
+  @NotNull
+  public Collection<VcsRepository> getRepositories() {
+    return mapping.values().stream().map(GitLabProject::getRepository).collect(Collectors.toList());
   }
 
   @Nullable

--- a/src/main/java/svnserver/repository/VcsRepositoryMapping.java
+++ b/src/main/java/svnserver/repository/VcsRepositoryMapping.java
@@ -13,7 +13,7 @@ import org.tmatesoft.svn.core.SVNException;
 import org.tmatesoft.svn.core.SVNURL;
 import svnserver.context.Shared;
 
-import java.io.IOException;
+import java.util.Collection;
 
 /**
  * Resolving repository by URL.
@@ -24,10 +24,6 @@ public interface VcsRepositoryMapping extends Shared {
   @Nullable
   RepositoryInfo getRepository(@NotNull SVNURL url) throws SVNException;
 
-  /**
-   * Update revision information in all mapped repositories.
-   *
-   * @throws IOException
-   */
-  void initRevisions() throws IOException, SVNException;
+  @NotNull
+  Collection<VcsRepository> getRepositories();
 }

--- a/src/main/java/svnserver/repository/mapping/RepositoryListMapping.java
+++ b/src/main/java/svnserver/repository/mapping/RepositoryListMapping.java
@@ -19,6 +19,7 @@ import svnserver.repository.VcsRepository;
 import svnserver.repository.VcsRepositoryMapping;
 
 import java.io.IOException;
+import java.util.Collection;
 import java.util.Map;
 import java.util.NavigableMap;
 import java.util.TreeMap;
@@ -53,14 +54,9 @@ public class RepositoryListMapping implements VcsRepositoryMapping {
   }
 
   @Override
-  public void initRevisions() throws IOException, SVNException {
-    new ConcurrentHashMap<>(mapping).forEachEntry(0, entry -> {
-      try {
-        entry.getValue().updateRevisions();
-      } catch (IOException | SVNException e) {
-        throw new RuntimeException(e);
-      }
-    });
+  @NotNull
+  public Collection<VcsRepository> getRepositories() {
+    return mapping.values();
   }
 
   @Nullable

--- a/src/main/java/svnserver/server/SvnServer.java
+++ b/src/main/java/svnserver/server/SvnServer.java
@@ -46,6 +46,7 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
 
 /**
  * Сервер для предоставления доступа к git-у через протокол subversion.
@@ -122,7 +123,20 @@ public class SvnServer extends Thread {
 
     repositoryMapping = config.getRepositoryMapping().create(context);
     context.add(VcsRepositoryMapping.class, repositoryMapping);
-    repositoryMapping.initRevisions();
+
+    final Consumer<VcsRepository> init = repository -> {
+      try {
+        repository.updateRevisions();
+      } catch (IOException | SVNException e) {
+        throw new RuntimeException(e);
+      }
+    };
+
+    if (config.canUseParallelIndexing()) {
+      repositoryMapping.getRepositories().parallelStream().forEach(init);
+    } else {
+      repositoryMapping.getRepositories().forEach(init);
+    }
 
     serverSocket = new ServerSocket();
     serverSocket.setReuseAddress(config.getReuseAddress());

